File: app/src/main/java/me/magnum/melonds/MelonEmulator.java
package me.magnum.melonds;

import java.nio.ByteBuffer;

import me.magnum.melonds.model.Input;

public final class MelonEmulator {
	public enum LoadResult {
		SUCCESS,
		SUCCESS_GBA_FAILED,
		NDS_FAILED
	}

	private MelonEmulator() {
	}

	public static native void setupEmulator(String configDir);

	public static LoadResult loadRom(String romPath, String sramPath, boolean loadDirect, boolean loadGbaRom, String gbaRomPath, String gbaSramPath) {
		int loadResult = loadRomInternal(romPath, sramPath, loadDirect, loadGbaRom, gbaRomPath, gbaSramPath);

		switch (loadResult) {
			case 0:
				return LoadResult.SUCCESS;
			case 1:
				return LoadResult.SUCCESS_GBA_FAILED;
			case 2:
				return LoadResult.NDS_FAILED;
		}
		throw new RuntimeException("Unknown load result");
	}

	private static native int loadRomInternal(String romPath, String sramPath, boolean loadDirect, boolean loadGbaRom, String gbaRomPath, String gbaSramPath);

	public static native void startEmulation();

	public static native void copyFrameBuffer(ByteBuffer frameBufferDst);

	public static native int getFPS();

	public static native void pauseEmulation();

	public static native void resumeEmulation();

	public static native void stopEmulation();

	public static native void onScreenTouch(int x, int y);

	public static native void onScreenRelease();

	public static void onInputDown(Input input) {
		onKeyPress(input.getKeyCode());
	}

	public static void onInputUp(Input input) {
		onKeyRelease(input.getKeyCode());
	}

	private static native void onKeyPress(int key);

	private static native void onKeyRelease(int key);
}

File: app/src/main/java/me/magnum/melonds/MelonEmulator.kt
package me.magnum.melonds

import android.content.res.AssetManager
import android.net.Uri
import me.magnum.melonds.common.RetroAchievementsCallback
import me.magnum.melonds.common.camera.DSiCameraSource
import me.magnum.melonds.domain.model.Cheat
import me.magnum.melonds.domain.model.EmulatorConfiguration
import me.magnum.melonds.domain.model.Input
import me.magnum.melonds.domain.model.retroachievements.RASimpleAchievement
import me.magnum.melonds.ui.emulator.render.FrameRenderCallback
import me.magnum.melonds.ui.emulator.rewind.model.RewindSaveState
import me.magnum.melonds.ui.emulator.rewind.model.RewindWindow
import java.nio.ByteBuffer

object MelonEmulator {
    enum class LoadResult(val isTerminal: Boolean) {
        SUCCESS(false),
        SUCCESS_GBA_FAILED(false),
        NDS_FAILED(true),
        BIOS_FAILED(true)
    }

    enum class FirmwareLoadResult {
        SUCCESS,
        BIOS9_MISSING,
        BIOS9_BAD,
        BIOS7_MISSING,
        BIOS7_BAD,
        FIRMWARE_MISSING,
        FIRMWARE_BAD,
        FIRMWARE_NOT_BOOTABLE,
        DSI_BIOS9_MISSING,
        DSI_BIOS9_BAD,
        DSI_BIOS7_MISSING,
        DSI_BIOS7_BAD,
        DSI_NAND_MISSING,
        DSI_NAND_BAD
    }

    enum class GbaSlotType {
        NONE,
        GBA_ROM,
        MEMORY_EXPANSION,
    }

	external fun setupEmulator(
        emulatorConfiguration: EmulatorConfiguration,
        assetManager: AssetManager?,
        dsiCameraSource: DSiCameraSource?,
        retroAchievementsCallback: RetroAchievementsCallback,
        screenshotBuffer: ByteBuffer,
        glContext: Long,
    )

    external fun setupCheats(cheats: Array<Cheat>)

    external fun setupAchievements(achievements: Array<RASimpleAchievement>, richPresenceScript: String?)

    external fun unloadAchievements(achievements: Array<RASimpleAchievement>)

    external fun getRichPresenceStatus(): String?

	fun loadRom(romUri: Uri, sramUri: Uri, gbaSlotType: GbaSlotType, gbaRomUri: Uri?, gbaSramUri: Uri?): LoadResult {
        val loadResult = loadRomInternal(romUri.toString(), sramUri.toString(), gbaSlotType.ordinal, gbaRomUri?.toString(), gbaSramUri?.toString())
        return when (loadResult) {
            0 -> LoadResult.SUCCESS
            1 -> LoadResult.SUCCESS_GBA_FAILED
            2 -> LoadResult.NDS_FAILED
            3 -> LoadResult.BIOS_FAILED
            else -> throw RuntimeException("Unknown load result")
        }
    }

    fun bootFirmware(): FirmwareLoadResult {
        val loadResult = bootFirmwareInternal()
        return FirmwareLoadResult.entries[loadResult]
    }

    private external fun loadRomInternal(romPath: String, sramPath: String, gbaSlotType: Int, gbaRomPath: String?, gbaSramPath: String?): Int

    private external fun bootFirmwareInternal(): Int

	external fun startEmulation()

    external fun presentFrame(frameRenderCallback: FrameRenderCallback)

	external fun getFPS(): Int

	external fun pauseEmulation()

	external fun resumeEmulation()

    external fun resetEmulation()

	external fun stopEmulation()

    fun saveState(path: Uri): Boolean {
        return saveStateInternal(path.toString())
    }

    private external fun saveStateInternal(path: String): Boolean

    fun loadState(path: Uri): Boolean {
        return loadStateInternal(path.toString())
    }

    private external fun loadStateInternal(path: String): Boolean

    external fun loadRewindState(rewindSaveState: RewindSaveState): Boolean

    external fun getRewindWindow(): RewindWindow

	external fun onScreenTouch(x: Int, y: Int)

	external fun onScreenRelease()

	fun onInputDown(input: Input) {
        onKeyPress(input.keyCode)
    }

	fun onInputUp(input: Input) {
        onKeyRelease(input.keyCode)
    }

    private external fun onKeyPress(key: Int)

    private external fun onKeyRelease(key: Int)

    external fun setFastForwardEnabled(enabled: Boolean)

    external fun setMicrophoneEnabled(enabled: Boolean)

    external fun updateEmulatorConfiguration(emulatorConfiguration: EmulatorConfiguration)
}